### DISCUSSION TITLE
Full chat bubbles customisation

### DISCUSSION THREAD

**Discussion by demchenkoalex** [flyerhq/flutter_chat_ui#102] [Ideas & Requests] [answered]
As in the title. Probably the next thing we will implement 🙂

**Comment by Androrier** [comment]
Can you elaborate?
I wish to add the person's title to the chat bubble, for example:

[John Doe]
[Rocket scientist]
[Hello world]

**Comment by demchenkoalex** [comment]
My thoughts was to add functions to build every type of message (right now there is only custom) and on top of that to add a function to build a custom bubble around them. Message builders will get messages (that contain authors) and bubble builder will get an indicator whether it's sent or received, something like that

**Reply by Androrier** [reply]
Thanks for reply.
After further investigating, I found out that I can use the customMessageBuilder to achieve this.
I created a custom_text_message which is much alike like text_message but the title and text on metadata.
Also, I can add the timestamp to the widget this way.
I can share code if you wish.

**Reply by demchenkoalex** [reply]
it's a hack we didn't predicted but you indeed can do a custom text message. But we will still create custom builders for every message type :)

The only thing is you can't actually change bubble around, so this is what we are going to add too.

We are not going to include timestamps by default anyway, but I guess you can create a discussion in "Show & Tell" category and share some screenshots (if you can) and code snippets to help other folks, who want to achieve a similar design :)

**Reply by Androrier** [reply]
Hi again, thanks for reply.
Is it possible to apply "showName" field to customMessageBuilder too (now it's only for TextMessage..)

**Reply by demchenkoalex** [reply]
no, it is impossible

1. Not everyone has custom messages and logic will be broken (some names will be missing)
2. We don't know the content of the custom message, it might be something not suitable for the text above it

When you build a custom message though you have author inside, so you can build it yourself I guess

**Reply by Androrier** [reply]
How would I know if two messages from the same user are consecutive?
(In order to set it false at the second, third etc'..)

**Reply by demchenkoalex** [reply]
that's a tricky one, even if I give you this boolean, the names will be shown on `TextMessage`s anyway, what you're asking is a specific case that you want to "include" your custom message in our processing function, but it cannot be easily done - see above. Even if we add like a parameter to "include" custom messages, what if for some custom messages it is wanted and for others not? Too many confusions with this one IMO

**Reply by Androrier** [reply]
I understand.
In my case, there will be no TextMessage since they don't contains a timestamp (Such as Whatsapp bubbles..)
Furthermore, I want to add a title to each user so I could only do it with the customMessage.
As I see it, I have no other choice but to fork your project :(

**Reply by demchenkoalex** [reply]
Wouldn't it be easier just to add a custom builder for text messages?

**Reply by Androrier** [reply]
It will be. If I could extend TextMessage - it will solve everything.
How do I do it?

**Reply by Androrier** [reply]
Oh, you want me to implement it and request a pr?

**Reply by demchenkoalex** [reply]
I am thinking it should just be buildTextMessage (similar to Custom) and maybe additional props like showName or something when name should be rendered (returned from our until algorithm) and that's it. You can try to do it or I will be able to do it on weekend, busy at work :(

**Reply by Androrier** [reply]
Let me see what I can do for you (and for me of course)

**Reply by Androrier** [reply]
Ok, I implemented it, ready for PR

**Reply by Androrier** [reply]
Let me know what you think.
Check the PR's

**Reply by demchenkoalex** [reply]
Released in [v1.4.2](https://github.com/flyerhq/flutter_chat_ui/releases/tag/v1.4.2).

**Reply by demchenkoalex** [reply]
I changed a couple of things, params related to link preview are not needed, because they are controlled on user's side anyway and are accessible, and I have added a `messageWidth` to all custom builders (may be helpful to see available space). Also while at it added custom builders for file and image messages.

**Reply by itaishalom** [reply]
Awesome work, thank you.

**Comment by demchenkoalex** [comment, marked as answer]
Released in [v1.4.3](https://github.com/flyerhq/flutter_chat_ui/releases/tag/v1.4.3).
